### PR TITLE
Fix PeriodicVariables2

### PR DESCRIPTION
**Pull request by dmdlott**
NEW HANDLING OF PERIODIC SCCS IN DOALL AND GINO:
- We now break the loop carried dependence of these SCCs by introducing an absolute iteration counter into the loop and directly calculating the SCC value from the counter at each iteration.

HAS A SISTER PR FROM NOELLE, REQUIRES THE SISTER PR

### CODE REVIEW

File: src/core/doall/src/DOALL_chunking.cpp [hunk 1]
```diff
@@ -157,17 +159,23 @@ void DOALL::rewireLoopToIterateChunks(LoopContent *LDI, DOALLTask *task) {
         && "DOALL: PHINode in periodic variable SCC doesn't have exactly two entries!");
     auto taskPHI = cast<PHINode>(task->getCloneOfOriginalInstruction(phi));
 
+    // We have a crummy assumption that periodic vars have only two incoming
```
Comment (scampanoni) [inline src/core/doall/src/DOALL_chunking.cpp]: Please be consistent with how we write comments.
/*
 * ...
 * /

File: src/core/doall/src/DOALL_chunking.cpp [hunk 2]
```diff
@@ -157,17 +159,23 @@ void DOALL::rewireLoopToIterateChunks(LoopContent *LDI, DOALLTask *task) {
...
       loopBlock = 1;
     } else {
+      // if the phinode doesnt have the incoming value for the periodic var,
+      // something is wrong with the PVSCC analysis
       assert(phi->getIncomingValue(1) == initialValue
              && "DOALL: periodic variable SCC selected the wrong PHINode!");
       entryBlock = 1;
       loopBlock = 0;
     }
+    // get clones of PVSCC vals
```
Comment (scampanoni) [inline src/core/doall/src/DOALL_chunking.cpp]: What is PVSCC?

File: src/core/doall/src/DOALL_chunking.cpp [hunk 3]
```diff
@@ -177,14 +185,68 @@ void DOALL::rewireLoopToIterateChunks(LoopContent *LDI, DOALLTask *task) {
...
+                                                  iterChunkStep,
+                                                  iterIncrement);
+      getOrInjectIterCounter->addIncoming(iterSelect,
+                                          latchBuilder.GetInsertBlock());
+    }
+
+    assert((getOrInjectIterCounter != nullptr)
+           && "DOALL_chunking: no iterCounter\n");
+
+    /* DD: we choose to not do this and instead replace the whole PVSCC with a
```
Comment (scampanoni) [inline src/core/doall/src/DOALL_chunking.cpp]: Please remove "DD"

Comment (scampanoni) [reply src/core/doall/src/DOALL_chunking.cpp]: Also, please follow the comment pattern we have been using throughout the codebase

File: src/core/doall/src/DOALL_chunking.cpp [hunk 4]
```diff
@@ -177,14 +185,68 @@ void DOALL::rewireLoopToIterateChunks(LoopContent *LDI, DOALLTask *task) {
...
+    computation based on the iteration.
+     * DD: there are other options here but we kind of just arbitrarily select
```
Comment (scampanoni) [inline src/core/doall/src/DOALL_chunking.cpp]: Remove "DD"

Comment (scampanoni) [reply src/core/doall/src/DOALL_chunking.cpp]: Please be precise and avoid ambiguous sentences like "we kind of just..."

File: src/core/doall/src/DOALL_chunking.cpp [hunk 5]
```diff
@@ -195,56 +257,32 @@ void DOALL::rewireLoopToIterateChunks(LoopContent *LDI, DOALLTask *task) {
...
-        cast<SelectInst>(chunkPHI->getIncomingValueForBlock(taskLoopBlock))
-            ->getCondition();
-    auto nextValue = loopBuilder.CreateSelect(isChunkCompleted,
-                                              nextChunkValue,
-                                              taskLoopValue,
-                                              "nextValue");
-    taskPHI->setIncomingValueForBlock(taskLoopBlock, nextValue);
+    headerBuilder.SetInsertPoint(
+        &*headerBuilder.GetInsertBlock()->getFirstInsertionPt());
+    // If there's some problem in here, consider whether e.g. the initialValue
```
Comment (scampanoni) [inline src/core/doall/src/DOALL_chunking.cpp]: I didn't get this comment.

**Review by scampanoni — CHANGES_REQUESTED** [review]
The code looks correct (great job!), but the comments need to be improved.

**Comment by dmdlott** [comment]
@scampanoni bumping it on gino too

**Pull request merged**